core/commands/filestore.go

```diff
@@ -185,6 +186,107 @@ For ERROR entries the error will also be printed to stderr.
 	Type: filestore.ListRes{},
 }
 
+var fixFileStore = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "Verify objects in filestore and remove bad links",
+		LongDescription: `
+Verify objects in filestore and remove bad links.
+
+If one or more <obj> is specified only verify those specific objects,
+otherwise verify all objects.
+
+The output is:
+
+<action> <status> <hash> <size> <path> <offset>
+
+Where <action> is one of:
+remove:   link will be removed
+nop:      nothing to do with it
+
+Where <status> is one of:
+ok:       the block can be reconstructed
+changed:  the contents of the backing file have changed
+no-file:  the backing file could not be found
+error:    there was some other problem reading the file
+missing:  <obj> could not be found in the filestore
+ERROR:    internal error, most likely due to a corrupt database
+
+For ERROR entries the error will also be printed to stderr.
+`,
+	},
+	Arguments: []cmds.Argument{
+		cmds.StringArg("obj", false, true, "Cid of objects to verify."),
+	},
+	Options: []cmds.Option{
+		cmds.BoolOption(fileOrderOptionName, "verify the objects based on the order of the backing file"),
+	},
+	Run: func(req *cmds.Request, res cmds.ResponseEmitter, env cmds.Environment) error {
+		_, fs, err := getFilestore(env)
+		if err != nil {
+			return err
+		}
+		args := req.Arguments
+		if len(args) > 0 {
+			return listByArgs(res, fs, args)
+		}
+
+		fileOrder, _ := req.Options[fileOrderOptionName].(bool)
+		next, err := filestore.VerifyAll(fs, fileOrder)
+		if err != nil {
+			return err
+		}
+
+		for {
+			r := next()
+			if r == nil {
+				break
+			}
+			if err := res.Emit(r); err != nil {
+				return err
+			}
+			if (r.Status != filestore.StatusOk) && (r.Status != filestore.StatusOtherError) {
+				fs.FileManager().DeleteBlock(r.Key)
+			}
+		}
+
+		return nil
+	},
+	PostRun: cmds.PostRunMap{
+		cmds.CLI: func(res cmds.Response, re cmds.ResponseEmitter) error {
+			enc, err := cmdenv.GetCidEncoder(res.Request())
+			if err != nil {
+				return err
+			}
+
+			for {
+				v, err := res.Next()
+				if err != nil {
+					if err == io.EOF {
+						return nil
+					}
+					return err
+				}
+
+				list, ok := v.(*filestore.ListRes)
+				if !ok {
+					return e.TypeErr(list, v)
+				}
+
+				if list.Status == filestore.StatusOtherError {
+					fmt.Fprintf(os.Stderr, "%s\n", list.ErrorMsg)
+				}
+
+				if (list.Status != filestore.StatusOk) && (list.Status != filestore.StatusOtherError) {
```

Comment:
Let's add the action to the output. We may want to change the action we take in some cases.